### PR TITLE
More on inputs

### PR DESCRIPTION
**Pull request by chrisgoringe**
Hidden inputs, more on flexible input types, and a few other details.

### CODE REVIEW

File: essentials/custom_node_more_on_inputs.mdx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+containing one or more of `PROMPT`, `EXTRA_PNGINFO`, or `UNIQUE_ID`
+
+```python
+@classmethod
+def INPUT_TYPES(s):
+    return {
+        "required": {...},
+        "optional": {...},
+        "hidden": {
+            "prompt": "PROMPT",
```
Comment (DrJKL) [inline essentials/custom_node_more_on_inputs.mdx:20]: How would you feel about arranging these and the headings below in the same order?
Either alphabetical or by perceived importance?

File: essentials/custom_node_more_on_inputs.mdx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+### PROMPT
+`PROMPT` is the complete prompt sent by the client to the server. 
+See [the prompt object](./javascript_objects_and_hijacking#prompt) for a full description.
+
+### EXTRA_PNGINFO
+`EXTRA_PNGINFO` is a dictionary that will be copied into the metadata of any `.png` files saved. Custom nodes can store additional
+information in this dictionary for saving (or as a way to communicate with a downstream node).
+
+<Tip>Note that if Comfy is started with the `disable_metadata` option, this data won't be saved.</Tip>
```
Comment (DrJKL) [inline essentials/custom_node_more_on_inputs.mdx:39]:
```suggestion
<Tip>Note that if Comfy is started with the `--disable_metadata` option, this data won't be saved.</Tip>
```

Comment (DrJKL) [reply essentials/custom_node_more_on_inputs.mdx:39]: Also a good link target if there's a page with all the options listed.

File: essentials/custom_node_more_on_inputs.mdx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+`*` is used to indicate that this `optional` input can be connected to any source. 
+It's up to the node to make sense of the data that is passed.
+<Tip>A `required` input may not be type `*` {/* TODO check if this is still true */}</Tip>
+
+### Dynamically created inputs
+
+If inputs are dynamically created on the client side, they can't be defined in the Python source code.
+In order to access this data we need an `optional` dictionary that allows Comfy to pass data with 
+arbitrary names. Since the Comfy server
```
Comment (DrJKL) [inline essentials/custom_node_more_on_inputs.mdx:81]: I think this sentence ...

File: essentials/javascript_objects_and_hijacking.mdx [hunk 4]
```diff
@@ -209,3 +209,6 @@ Custom widget types can be added by providing a `getCustomWidgets` method in you
 Widgets can also be linked - the built in behavior of `seed` and `control_after_generate`, for example. 
 A linked widget has `.type = 'base_widget_type:base_widget_name'`; so `control_after_generate` may have
 type `int:seed`.
+
+## Prompt
+
```
Comment (DrJKL) [inline essentials/javascript_objects_and_hijacking.mdx:214]: Empty heading?

File: essentials/custom_node_more_on_inputs.mdx (hunk 1 above)
Comment (chrisgoringe) [reply essentials/custom_node_more_on_inputs.mdx:20]: Yup. Pushing update in a minute with this and other tweaks.

File: essentials/custom_node_more_on_inputs.mdx (hunk 2 above)
Comment (chrisgoringe) [reply essentials/custom_node_more_on_inputs.mdx:39]: Added this as a TODO comment for later searching!

File: essentials/javascript_objects_and_hijacking.mdx (hunk 4 above)
Comment (chrisgoringe) [reply essentials/javascript_objects_and_hijacking.mdx:214]: Working on this right now - happy to hold this PR back to wait for the next one

File: essentials/custom_node_more_on_inputs.mdx (hunk 3 above)
Comment (chrisgoringe) [reply essentials/custom_node_more_on_inputs.mdx:81]: ah...

**Comment by chrisgoringe** [comment]
I've added the empty stuff in PR21, and merged this branch into that one, so this PR can be closed.

**Pull request closed without merging**